juju/juju
# k8s bootstraping: make sure that namespace is of lower case

## Description of change

if no name is given during k8s bootstrapping juju will bootstrap a controller with the following name:
`<cloudname>-<regionname>`.

For k8 we can have the problem because we use the controller name as a namespace as well.
Regions can contain capitalized characters which are not possible for namespaces.

This would lead to the following:
```
ERROR failed to bootstrap model: creating controller stack for controller: creating namespace for controller stack: Namespace "controller-internal-k8s-RegionOne" is invalid: metadata.name: Invalid value: "controller-internal-k8s-RegionOne": a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')
```

The change forces controller names to be of lowercase
## QA steps
- unit test are added
bootstrap output
```shell
[17:10:10] nam:~ $ juju bootstrap microk8s CAPS        
Creating Juju controller "CAPS" on microk8s/localhost
Creating k8s resources for controller "controller-caps"
Starting controller pod
Bootstrap agent now started
Contacting Juju controller at 10.152.183.193 to verify accessibility...

Bootstrap complete, controller "CAPS" now is available in namespace "controller-caps"

Now you can run
	juju add-model <model-name>
to create a new model to deploy k8s workloads.

```
namespaces output
```shell 
[17:21:48] nam:~ $ mkc get namespaces                     
NAME                      STATUS   AGE
controller-caps           Active   38s
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1842991

## Update 

This change should force all bootstrapped controller to be lowercase.
Below output examples using CAPS in names.
```
[16:52:45] nam:juju git:(fix-k8s-bootstrapping*) $ juju bootstrap microk8s CAPITILAdsadsa
Creating Juju controller "capitiladsadsa" on microk8s/localhost
Creating k8s resources for controller "controller-capitiladsadsa"
Starting controller pod
Bootstrap agent now started
Contacting Juju controller at 10.152.183.75 to verify accessibility...

Bootstrap complete, controller "capitiladsadsa" now is available in namespace "controller-capitiladsadsa"

Now you can run
        juju add-model <model-name>
to create a new model to deploy k8s workloads.

[17:00:52] nam:juju git:(fix-k8s-bootstrapping) $ mkc get namespaces
NAME                        STATUS   AGE
controller-capitiladsadsa   Active   6m59s
```

```
[17:02:19] nam:juju git:(fix-k8s-bootstrapping) $ juju bootstrap localhost/Localhost
Creating Juju controller "localhost-localhost" on localhost/localhost
Looking for packaged Juju agent version 2.7-beta1 for amd64
...
```

```
[17:01:11] nam:juju git:(fix-k8s-bootstrapping) $ juju bootstrap localhost/Localhost
ERROR controller "localhost-localhost" already exists

```

To make sure that I only changed the controllerName and there is no things going around unexpected, I took a look at my cloud-config. 
```
clouds:

  localhost:
    type: lxd
    auth-types: [certificate]
    regions:
      localhost: {}
      Localhost: {}
```
